opencog/reasoning/RuleEngine/rule-engine-src/pln/PLNCommons.cc

```diff
@@ -0,0 +1,194 @@
+/*
+ * Commons.cc
+ *
+ *  Created on: 9 Oct, 2014
+ *      Author: misgana
+ */
+
+#include "PLNCommons.h"
+
+#include <opencog/util/macros.h>
+#include <opencog/guile/SchemeSmob.h>
+
+PLNCommons::PLNCommons(AtomSpace * as) :
+		as_(as) {
+
+}
+
+PLNCommons::~PLNCommons() {
+
+}
+Handle PLNCommons::create_quoted(Handle himplicant) {
+	Handle hquoted;
+	if (LinkCast(himplicant)) {
+		HandleSeq hs = as_->getOutgoing(himplicant);
+		HandleSeq hs_new;
+		for (Handle h : hs)
+			hs_new.push_back(create_quoted(h));
+		hquoted = as_->addLink(as_->getType(himplicant), hs_new);
+	} else if (as_->getType(himplicant) == VARIABLE_NODE)
+		hquoted = as_->addLink(QUOTE_LINK, himplicant);
+	else
+		hquoted = himplicant;
+	return hquoted;
+}
+
+Handle PLNCommons::create_bindLink(Handle himplicant, bool vnode_is_typedv)
+		throw (opencog::InvalidParamException) {
+	if (!LinkCast(himplicant)) {
+		throw InvalidParamException(TRACE_INFO, "Input must be a link type ");
+	} //xxx why?
+
+	//if(vnode_is_typedv)
+	himplicant = create_with_unique_var(himplicant);
+
+	HandleSeq variable_nodes = get_nodes(himplicant, vector<Type> {
```

Comment:
I almost hate to mention this, as perhaps its low priority, but .. the pattern matcher does have lower-level API's than bind-link.   If you already know what the variable are, and what the clauses are, you can pass them directly to the pattern matcher.

When you give the pattern matcher a bind link, the very first thing it does is to extract the variables, and reorganize everything into C++ HandleSeq's. So, if you already have handleSeq's, there is not much point to pack them into a bind-link, which the pattern matcher must then unpack. 

This is low priority, though, because there is no harm to the pack/unpack step, and it does give the pattern matcher a chance to look for badly-formed patterns, and throw errors in such cases ...